### PR TITLE
feat(perftool): add perftool

### PR DESCRIPTION
**Pull request by akhdrv**
:baby_chick: Download canary assets:

[colors--canary.268.3782888639.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.268.3782888639.xml)  
[color_metro_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.268.3782888639.swift)  
[color_metro_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.268.3782888639.kt)  
[color_metro_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.268.3782888639.ts)  
[color_sbermarket_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.268.3782888639.swift)  
[color_sbermarket_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.268.3782888639.kt)  
[color_sbermarket_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.268.3782888639.ts)  
[color_sberprime_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.268.3782888639.swift)  
[color_sberprime_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.268.3782888639.kt)  
[color_sberprime_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.268.3782888639.ts)  
[color_selgros_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.268.3782888639.swift)  
[color_selgros_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.268.3782888639.kt)  
[color_selgros_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.268.3782888639.ts)  
[color_smbusiness_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.268.3782888639.swift)  
[color_smbusiness_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.268.3782888639.kt)  
[color_smbusiness_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.268.3782888639.ts)  
[PlasmaTokensColor--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.268.3782888639.swift)  
[typo_mage_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.268.3782888639.swift)  
[typo_mage_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.268.3782888639.kt)  
[typo_mage_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.268.3782888639.ts)  
[typo_plasma_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.268.3782888639.swift)  
[typo_plasma_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.268.3782888639.kt)  
[typo_plasma_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.268.3782888639.ts)  
[typo_ruler_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.268.3782888639.swift)  
[typo_ruler_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.268.3782888639.kt)  
[typo_ruler_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.268.3782888639.ts)  
[typo_sage_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.268.3782888639.swift)  
[typo_sage_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.268.3782888639.kt)  
[typo_sage_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.268.3782888639.ts)  
[typo_sbermarket_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.268.3782888639.swift)  
[typo_sbermarket_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.268.3782888639.kt)  
[typo_sbermarket_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.268.3782888639.ts)  
[typo_soulmate_ios-swift--canary.268.3782888639.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.268.3782888639.swift)  
[typo_soulmate_kotlin--canary.268.3782888639.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.268.3782888639.kt)  
[typo_soulmate_react-native--canary.268.3782888639.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.268.3782888639.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-temple@1.121.0-canary.268.3782888639.0
  npm install @salutejs/plasma-ui@1.154.0-canary.268.3782888639.0
  npm install @salutejs/perftool@0.2.0-canary.268.3782888639.0
  # or 
  yarn add @salutejs/plasma-temple@1.121.0-canary.268.3782888639.0
  yarn add @salutejs/plasma-ui@1.154.0-canary.268.3782888639.0
  yarn add @salutejs/perftool@0.2.0-canary.268.3782888639.0
  ```
</details>

### CODE REVIEW

File: packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx [hunk 1]
```diff
@@ -0,0 +1,233 @@
+import React from 'react';
+import styled from 'styled-components/macro';
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx:2]: зачем /macro?
```js
import styled from 'styled-components' 
```

File: packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    TimePickerStep: 1,
+    TimePickerVisibleItems: 5,
+    DatePickerNativeControl: false,
+    TimePickerNativeControl: false,
+};
+
+export const Default = () => {
+    const args = defaultComponentArgs;
+    const [value, setValue] = React.useState(parseDateTime(args.initialValue));
+    const min = React.useMemo(() => parseDateTime(args.minDate), [args.minDate]);
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx:49]: автоимпорт в vscode всегда сам добавляет
```js
import { useMemo } from 'react'
```

File: packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+            onChange={setValue}
+            daysAriaLabel="день"
+            monthsAriaLabel="месяц"
+            yearsAriaLabel="год"
+            {...rest}
+        />
+    );
+};
+
+const timePickerProps = {
```
Comment (KateKate) [inline packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx:183]: до этого были args - datePickerArgs, defaultComponentArgs. а тут стали props. давай один подход. кажется props понятнее

File: utils/perftool/src/build/collect.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            modules.push(entryPath);
+        }
+    }
+
+    return modules;
+}
+
+const PICK_RULE_METHODS: Record<ExportPickRule, (fileContents: string) => string[]> = {
+    named(fileContents) {
+        const regex = /export\s+(?:const|var|let|function)\s+?([\w$_][\w\d$_]*)/g;
```
Comment (KateKate) [inline utils/perftool/src/build/collect.ts:54]: а ты не хочешь тут через AST? https://babeljs.io/docs/en/babel-parser#output

File: utils/perftool/src/build/entrypoint.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            return subjects.map(({ id }) => `{ id: '${id}', Component: ${id} },`);
+        })
+        .flat();
+    global.console.log(clientTestSubjects);
+
+    const contents = await fsPromises.readFile(entrypointPath, { encoding: 'utf-8' });
+
+    const formattedContents = contents
+        .replace('// <IMPORT_MARK>', formatLines(imports))
+        .replace('// <TEST_SUBJECT_MARK>', formatLines(clientTestSubjects));
```
Comment (KateKate) [inline utils/perftool/src/build/entrypoint.ts:37]: а давай комментарии что означает <IMPORT_MARK> и <TEST_SUBJECT_MARK>

Comment (KateKate) [inline utils/perftool/src/build/entrypoint.ts:37]: а давай комментарии что означает <IMPORT_MARK> и <TEST_SUBJECT_MARK>

File: utils/perftool/src/client/index.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { Config } from '../config';
+
+import { Task } from './measurement/types';
+import { runTask, Subject } from './measurement/runner';
+import { sendResults } from './sendResults';
+import { resolveCurrentSessionParams } from './sessionParams';
+
+type CreatePerfToolClientParams<T extends Task<any, any>[]> = {
```
Comment (KateKate) [inline utils/perftool/src/client/index.ts:8]: Task<unknown, unknown>

File: utils/perftool/src/client/measurement/index.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import render from './tasks/render';
+import rerender from './tasks/rerender';
```
Comment (KateKate) [inline utils/perftool/src/client/measurement/index.ts:2]: давай именованные import и export, без дефолтных

File: utils/perftool/src/client/measurement/runner.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
```
Comment (KateKate) [inline utils/perftool/src/client/measurement/runner.ts:1]:
```suggestion
```

File: utils/perftool/src/client/measurement/runner.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+import { Config, getTaskConfig } from '../../config';
+import BaseError from '../../utils/baseError';
```
Comment (KateKate) [inline utils/perftool/src/client/measurement/runner.ts:4]: давай именованные импорты/экспорты везде. в ide/автоимпорты с ними проще + default ведет себя по-другому

https://jakearchibald.com/2021/export-default-thing-vs-thing-as-default/#:~:text=by%20value%20again.-,But%E2%80%A6%20why%3F,-It%20isn%27t%20just

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-268/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-268/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-268/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-268/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-268/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-268/temple-storybook/

File: utils/perftool/src/client/sessionParams.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+        return { runId: runId as string, tests };
+    }
+
+    if (Array.isArray(window.tests)) {
+        return transformResult(window.tests);
+    }
+
+    return new Promise((resolve, reject) => {
+        window.tests = {
```
Comment (KateKate) [inline utils/perftool/src/client/sessionParams.ts:52]: а тут так и останется в `window`?

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-268/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-268/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-268/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-268/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-268/temple-storybook/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-268/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-268/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-268/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-268/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-268/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-268/temple-storybook/

File: packages/plasma-ui/perftool.config.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    include: ['src/**/Pickers.perftest.tsx'],
+    modifyWebpackConfig: function (config) {
+        const babelLoaderOpts = config.module.rules.find(
+            (rule) => typeof rule === 'object' && rule.loader === 'babel-loader',
+        );
+
+        if (typeof babelLoaderOpts == 'object' && typeof babelLoaderOpts?.options === 'object') {
+            babelLoaderOpts?.options?.plugins?.push('babel-plugin-styled-components');
+
+            // ['@babel/plugin-proposal-decorators', { legacy: true }], // to project config
```
Comment (RPG59) [inline packages/plasma-ui/perftool.config.ts:15]: это лучше удалить

File: utils/perftool/src/client/sendResults.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import { RunTaskResult } from './measurement/runner';
+import { Task } from './measurement/types';
+
+export type SendResultsParams<T extends Task<any, any>> = {
+    runId: string;
+    results: RunTaskResult<T>[];
+};
+
+export async function sendResults<T extends Task<any, any>>(results: SendResultsParams<T>): Promise<void> {
+    global.console.log(results);
```
Comment (akhdrv) [inline utils/perftool/src/client/sendResults.ts:12]: убрать, добавить логирования на клиент

File: utils/perftool/src/utils/ErrorBoundary.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+
+import BaseError from './baseError';
+
+type ErrorBoundaryProps = {
+    onError: (error: BaseError) => void;
+    children: React.ReactNode;
+};
+
+class ErrorBoundary extends React.Component<ErrorBoundaryProps> {
```
Comment (akhdrv) [inline utils/perftool/src/utils/ErrorBoundary.tsx:10]: впилить

File: utils/perftool/src/build/index.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+import { buildDirectory, Config, distDirectory, getWebpackConfig, sourceDirectory } from '../config';
+import { CliConfig } from '../config/common';
+import { debug, error, info, warn } from '../utils/logger';
+
+import { TestModule } from './collect';
+import { modifyEntrypoint } from './entrypoint';
+
+const dirname = path.dirname(fileURLToPath(import.meta.url));
+
+// // TODO memfs unionfs
```
Comment (RPG59) [inline utils/perftool/src/build/index.ts:17]: давай создадим задачку на это, а коммент удалим

File: packages/plasma-ui/perftool.config.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import type { Config } from '@salutejs/perftool';
+
+const config: Config = {
+    jobs: 3,
+    retries: 10,
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply packages/plasma-ui/perftool.config.ts:5]: нужно скорее всего больше

File: packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+    minDate: '01.01.1975 01:15:29',
+    maxDate: '31.12.1985 12:45:50',
+    optionsYears: true,
+    optionsMonths: true,
+    optionsDays: true,
+    optionsShortMonthName: false,
+    optionsHours: true,
+    optionsMinutes: true,
+    optionsSeconds: true,
+    disabled: false,
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply packages/plasma-ui/src/components/Pickers/Pickers.perftest.tsx:33]: по некоторым причинам проще выносить компоненты. https://mm.sberdevices.ru/sberdevices/pl/u1epj8mibigezksyd3gqf9xqtc

а так по факту я и взял дефолтные опции из stories. вполне возможно что нужно меньше, не разбирался

File: utils/perftool/.nvmrc [hunk 17]
```diff
@@ -0,0 +1 @@
+16.15.0
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/.nvmrc:1]: сама тулза будет запускаться под 16 нодой, но пока вы не перешли на нее, придется использовать 6 npm. ужасно =), но в обратно на cjs не оч хочется переходить, там некоторые пакетики надо даунгрейдить и прочие приколы

File: utils/perftool/README.md [hunk 18]
```diff
@@ -0,0 +1,4 @@
+# Performance measurement tool for React components
+
+Node 16
+TODO
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/README.md:4]: =)

File: utils/perftool/package.json [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        "react": "18.2.0",
+        "react-dom": "18.2.0",
+        "typescript": "4.9.3"
+    },
+    "scripts": {
+        "prepare": "npm run build",
+        "prebuild": "npm run clean",
+        "build": "tsc",
+        "clean": "rm -rf lib",
+        "prepack": "npm run build"
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/package.json:56]: ну хз, когда вызывал npm pack, один раз вызывалось

File: utils/perftool/package.json [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    },
+    "publishConfig": {
+        "access": "public"
+    },
+    "files": [
+        "lib",
+        "src"
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/package.json:63]: нужно для сборки клиента измерялки. так было проще в какой-то момент, но честно уже не помню почему. добавлю в беклог

File: utils/perftool/src/build/index.ts [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+import fs from 'fs';
+
+import { buildDirectory, Config, distDirectory, getWebpackConfig, sourceDirectory } from '../config';
+import { CliConfig } from '../config/common';
+import { debug, error, info, warn } from '../utils/logger';
+
+import { TestModule } from './collect';
+import { modifyEntrypoint } from './entrypoint';
+
+const dirname = path.dirname(fileURLToPath(import.meta.url));
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/src/build/index.ts:15]: да

File: utils/perftool/package.json (hunk 20 above)
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/package.json:63]: вообще есть идея ts-node использовать и отказаться от lib соответственно

File: utils/perftool/src/client/sessionParams.ts (hunk 10 above)
Comment (akhdrv) [reply utils/perftool/src/client/sessionParams.ts:52]: да, все взаимодействие puppeteer'ом идет через глобальный объект

File: utils/perftool/src/client/measurement/runner.ts (hunk 8 above)
Comment (akhdrv) [reply utils/perftool/src/client/measurement/runner.ts:1]: оставляем поддержку 16 ноды

File: utils/perftool/src/build/collect.ts (hunk 4 above)
Comment (akhdrv) [reply utils/perftool/src/build/collect.ts:28]: были риски потратить на это много времени, плюс конфиг надо менять. добавлю в задачу на доработку

File: utils/perftool/src/server/index.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+    if (config.logLevel === 'verbose') {
+        app.use(
+            morgan('dev', {
+                skip: (req) => req.path === '/',
+            }),
+        );
+    }
+
+    app.use('/result', express.json());
```
Comment (akhdrv) [inline utils/perftool/src/server/index.ts:36]: TODO puppeteer page.exposeFunction и выпилить получение результатов через сервер

File: utils/perftool/package.json (hunk 20 above)
[the remark this answers is not in the document]
Comment (KateKate) [reply utils/perftool/package.json:63]: Артем, запиши плиз это тоже в беклог

File: utils/perftool/src/client/sessionParams.ts (hunk 10 above)
Comment (KateKate) [reply utils/perftool/src/client/sessionParams.ts:52]: а добавь комментарий про это, то не очевидно

File: utils/perftool/src/build/index.ts (hunk 14 above)
Comment (KateKate) [reply utils/perftool/src/build/index.ts:17]: задачку надо и TODO оставим. Ничего плохого не вижу в  TODO в коде. Наоборот полезно, когда код читаешь. Понимаешь, что может уже тут планируется то, про что ты подумал

File: utils/perftool/src/compare/process.ts [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+    for (const [currentMetricId, currentMetricResult] of Object.entries(restReport)) {
+        const metric = metricIdToMetricMap[currentMetricId];
+
+        assert(metric);
+
+        const previousMetricResult = isStatsMap(previous) ? previous[currentMetricId] : undefined;
+        const isCurrentMetricResultWithError = Array.isArray(currentMetricResult);
+        const isPreviousMetricResultWithError = Array.isArray(previousMetricResult);
+
+        if (isCurrentMetricResultWithError !== isPreviousMetricResultWithError) {
```
Comment (KateKate) [inline utils/perftool/src/compare/process.ts:127]: вот тут не осознала условие. почему проверка на массив? и либо всегда два массива, либо нет

File: utils/perftool/src/config/webpack.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import webpack, { Configuration as WebpackConfig } from 'webpack';
+
+import { debug } from '../utils/logger';
+
+import { Config } from './common';
+
+const defaultConfig: WebpackConfig = {
```
Comment (KateKate) [inline utils/perftool/src/config/webpack.ts:8]: давай в беклог - подумать заменить на parcel/vite. возможно тут Webpack избыточен

File: utils/perftool/src/controller/executor.ts [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+    async execute(tests: Test[]): Promise<RunTaskResult<T[number]>[] | Error> {
+        debug('[executor]', 'running tests', tests);
+
+        const page = await (this.browserInstance as Browser).newPage();
+        const runId = String(this.currentRunId++);
+
+        const results = new Deferred<RunTaskResult<T[number]>[]>();
+        this.currentTasks.set(runId, results);
+
+        await page.goto(`http://localhost:${this.port}/?runId=${runId}`);
```
Comment (KateKate) [inline utils/perftool/src/controller/executor.ts:114]: localhost - норм?

File: utils/perftool/src/controller/index.ts [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+
+type RunTestsParams<T extends [...Task<any, any>[]]> = {
+    config: Config;
+    notifier: Emitter<ResultEvent<T>>;
+    port: number;
+    tasks: T;
+    testModules: TestModule[];
+};
+
+export async function* runTests<T extends [...Task<any, any>[]]>({
```
Comment (KateKate) [inline utils/perftool/src/controller/index.ts:21]: any => unknown

File: utils/perftool/src/controller/planner.ts [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+
+    constructor(config: Config, tasks: T, testModules: TestModule[]) {
+        this.config = config;
+        this.tasks = tasks;
+        this.testModules = testModules;
+    }
+
+    *plan(): Generator<Test[]> {
+        const idempotentTasks = this.tasks.filter(({ isIdempotent }) => isIdempotent);
+        const nonIdempotentTasks = this.tasks.filter(({ isIdempotent }) => !isIdempotent);
```
Comment (KateKate) [inline utils/perftool/src/controller/planner.ts:43]: idempotentTasks - это одни и те же тесты, просто несколько повторений?

**Review by KateKate — APPROVED** [review]
Учитывая комменты + занесение в беклог TODO

File: utils/perftool/src/build/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (Yeti-or) [reply utils/perftool/src/build/index.ts:15]: а зачем тогда так сложно?

File: utils/perftool/package.json (hunk 19 above)
[the remark this answers is not in the document]
Comment (Yeti-or) [reply utils/perftool/package.json:56]: У тебя публикация (  npm publish  )  вызывается сама в релизе:

оно будет сначала дергать prepare потом pack

```
 perftool % npm publish -dry-run

> @salutejs/perftool@0.1.0 prepare .
> npm run build

> @salutejs/perftool@0.1.0 prepack .
> npm run build

```

File: utils/perftool/src/build/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/src/build/index.ts:15]: __dirname нету в esm

File: utils/perftool/package.json (hunk 19 above)
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/package.json:56]: уберу

File: utils/perftool/src/build/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (Yeti-or) [reply utils/perftool/src/build/index.ts:15]: ааа

**Review by Yeti-or — APPROVED** [review]

File: utils/perftool/src/config/webpack.ts (hunk 24 above)
Comment (akhdrv) [reply utils/perftool/src/config/webpack.ts:8]: не согласен что webpack избыточен, нам нужно собирать клиент в соответствие с тем, как собирается основное приложение. проекты в основном собираются вебпаком и у каждого могут быть свои нюансы соответственно

File: utils/perftool/src/compare/process.ts (hunk 23 above)
Comment (akhdrv) [reply utils/perftool/src/compare/process.ts:127]: это проверка на кортеж. там либо number, либо [number, number]

File: utils/perftool/tsconfig.json [hunk 28]
```diff
@@ -0,0 +1,42 @@
+{
+    "compilerOptions": {
+        "target": "ES2021",
+        "module": "esnext",
+        "lib": ["dom", "dom.iterable", "ES2021"],
+        "jsx": "react",
+        "declaration": true,
+        "declarationMap": true,
+        "sourceMap": true,
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/tsconfig.json:9]: вообще нет, нужно вырубить

File: utils/perftool/src/utils/waitForIdle.ts [hunk 29]
```diff
@@ -0,0 +1,13 @@
+function waitForIdle(): Promise<void> {
+    const idle = (window as any).requestIdleCallback;
+
+    if (typeof idle !== 'function') {
+        return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/src/utils/waitForIdle.ts:5]: я бы тут инвариант вообще поставил. все современные хромы поддерживают requestIdleCallback

File: utils/perftool/src/config/webpack.ts (hunk 24 above)
Comment (KateKate) [reply utils/perftool/src/config/webpack.ts:8]: тут же просто компоненты проверяем, то есть только React надо. какие нюансы. next кстати на [turbopack](https://turbo.build/pack/docs/why-turbopack) переезжает

File: utils/perftool/src/utils/waitForIdle.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (KateKate) [reply utils/perftool/src/utils/waitForIdle.ts:5]: safari еще не поддерживает - офигеть https://caniuse.com/?search=requestIdleCallback

File: utils/perftool/src/compare/process.ts (hunk 23 above)
Comment (KateKate) [reply utils/perftool/src/compare/process.ts:127]: я по логике не понимаю почему такая проверка. кажется, что всегда должно быть либо массив/кортеж или число

Comment (akhdrv) [reply utils/perftool/src/compare/process.ts:127]: сейчас да. но что будет в будущем никто не знает, поэтому это соломка для возможных изменений в будущих версиях

File: utils/perftool/src/config/webpack.ts (hunk 24 above)
Comment (akhdrv) [reply utils/perftool/src/config/webpack.ts:8]: нюансы — это вебпак плагины и лоудеры. пока вебпака больше, чем всех остальных сборщиков вместе взятых, это нельзя не учитывать. и хотелось бы, чтобы тулза была максимально независима от нашего стека. нексту с его турбопаком это не помешает. поэтому о таком подумать стоит разве что через несколько лет

File: utils/perftool/src/controller/executor.ts (hunk 25 above)
Comment (akhdrv) [reply utils/perftool/src/controller/executor.ts:114]: а что с ним может быть не норм?

File: utils/perftool/src/controller/index.ts (hunk 26 above)
Comment (akhdrv) [reply utils/perftool/src/controller/index.ts:21]: там ограничение, unknown не пойдет. сделаю дефолтные аргументы, чтобы везде any не писать

File: utils/perftool/src/controller/planner.ts (hunk 27 above)
Comment (akhdrv) [reply utils/perftool/src/controller/planner.ts:43]: да, это таски, результаты которых могут меняться, соответственно повторений должно быть несколько, и по ним нужно считать статистику

File: utils/perftool/src/utils/waitForIdle.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (akhdrv) [reply utils/perftool/src/utils/waitForIdle.ts:5]: сафари к счастью не используем для измерений

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-268/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-268/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-268/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-268/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-268/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-268/temple-storybook/

**Pull request merged**